README.md
# DebounceEvent

Simple push button and toggle switch debounce library for AVR and ESP8266 platforms under the Arduino framework that reports number of clicks and length

[![version](https://img.shields.io/badge/version-2.0.5-brightgreen.svg)](CHANGELOG.md)
[![license](https://img.shields.io/badge/license-LGPL--3.0-orange.svg)](LICENSE)
<br />
[![donate](https://img.shields.io/badge/donate-PayPal-blue.svg)](https://www.paypal.com/cgi-bin/webscr?cmd=_donations&business=xose%2eperez%40gmail%2ecom&lc=US&no_note=0&currency_code=EUR&bn=PP%2dDonationsBF%3abtn_donate_LG%2egif%3aNonHostedGuest)
[![twitter](https://img.shields.io/twitter/follow/xoseperez.svg?style=social)](https://twitter.com/intent/follow?screen_name=xoseperez)

## Features

The main features of the DebounceEvent library are:

* Supports push buttons and toggle switches
* Reports # of clicks and length (in millis) of the last click
* Allows to define internal pull-ups
* Isolation between buttons
* Create modular code using callbacks

## Usage

See examples.

## License

Copyright (C) 2015-2018 by Xose Pérez <xose dot perez at gmail dot com>

The DebounceEvent library is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

The DebounceEvent library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with the DebounceEvent library.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG.md
# DebounceEvent change log

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [2.0.5] 2019-02-07
### Fixed
- Do digitalRead after setting pin mode (thanks to @mcspr)

## [2.0.4] 2018-09-09
### Fixed
- Examples file names and keywords compatible with Arduino IDE (thanks to @per1234)

### Added
- Suport for AVR boards

## [2.0.3] 2018-08-29
### Fixed
- Fixed initial status for switches (thanks to @BauerPh)

## [2.0.2] 2018-06-23
### Changed
- Moved license to LGPL-3.0

## [2.0.1] 2017-07-23
### Added
- Added the option to specify repeat delay in constructor (set it to 0 to disable double click but get immediate responses)

## [2.0.0] 2017-01-25
### Changed
- Reporting # of clicks and length instead of predefined events
- Callback signature changed to report click count and length

### Removed
- EVENT_SINGLE_CLICK, EVENT_DOUBLE_CLICK, EVENT_LONG_CLICK

## [1.1.0] 2017-01-22
### Added
- Added support for toggle switches

### Fixed
- Fix bug with callback initialization

### Changed
- Changed constructor signature

## [1.0.0] 2016-12-29
- Initial commit as independent library


src/DebounceEvent.cpp
/*

  Debounce buttons and trigger events
  Copyright (C) 2015-2018 by Xose Pérez <xose dot perez at gmail dot com>

  The DebounceEvent library is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  The DebounceEvent library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with the DebounceEvent library.  If not, see <http://www.gnu.org/licenses/>.


*/

#include <Arduino.h>
#include "DebounceEvent.h"

DebounceEvent::DebounceEvent(uint8_t pin, DEBOUNCE_EVENT_CALLBACK_SIGNATURE, uint8_t mode, unsigned long delay, unsigned long repeat) {
    this->callback = callback;
    _init(pin, mode, delay, repeat);
}

DebounceEvent::DebounceEvent(uint8_t pin, uint8_t mode, unsigned long delay, unsigned long repeat) {
    this->callback = NULL;
    _init(pin, mode, delay, repeat);
}


void DebounceEvent::_init(uint8_t pin, uint8_t mode, unsigned long delay, unsigned long repeat) {

    // store configuration
    _pin = pin;
    _mode = mode & 0x01;
    _defaultStatus = ((mode & BUTTON_DEFAULT_HIGH) > 0);
    _delay = delay;
    _repeat = repeat;

    // set up button
    #if ESP8266
    if (_pin == 16) {
        if (_defaultStatus) {
            pinMode(_pin, INPUT);
        } else {
            pinMode(_pin, INPUT_PULLDOWN_16);
        }
    } else {
    #endif // ESP8266
        if ((mode & BUTTON_SET_PULLUP) > 0) {
            pinMode(_pin, INPUT_PULLUP);
        } else {
            pinMode(_pin, INPUT);
        }
    #if ESP8266
    }
    #endif // ESP8266

    _status = (_mode == BUTTON_SWITCH) ? digitalRead(_pin) : _defaultStatus;

}

unsigned char DebounceEvent::loop() {

    unsigned char event = EVENT_NONE;

    if (digitalRead(_pin) != _status) {

        // Debounce
        unsigned long start = millis();
        while (millis() - start < _delay) delay(1);

        if (digitalRead(_pin) != _status) {

            _status = !_status;

            if (_mode == BUTTON_SWITCH) {

                event = EVENT_CHANGED;

            } else {

                // released
                if (_status == _defaultStatus) {

                    _event_length = millis() - _event_start;
                    _ready = true;

                // pressed
                } else {

                    event = EVENT_PRESSED;
                    _event_start = millis();
                    _event_length = 0;
                    if (_reset_count) {
                        _event_count = 1;
                        _reset_count = false;
                    } else {
                        ++_event_count;
                    }
                    _ready = false;

                }

            }

        }
    }

    if (_ready && (millis() - _event_start > _repeat)) {
        _ready = false;
        _reset_count = true;
        event = EVENT_RELEASED;
    }

    if (event != EVENT_NONE) {
        if (this->callback) this->callback(_pin, event, _event_count, _event_length);
    }

    return event;

}


examples/debounceevent_basic/debouceevent_basic.cpp
/*

  Debounce buttons and trigger events
  Copyright (C) 2015-2017 by Xose Pérez <xose dot perez at gmail dot com>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <Arduino.h>
#include <DebounceEvent.h>

#define BUTTON_PIN              0

#define CUSTOM_DEBOUNCE_DELAY   50

// Time the library waits for a second (or more) clicks
// Set to 0 to disable double clicks but get a faster response
#define CUSTOM_REPEAT_DELAY     500

DebounceEvent * button;

void setup() {
    Serial.begin(115200);
    Serial.println();
    Serial.println();
    button = new DebounceEvent(BUTTON_PIN, BUTTON_PUSHBUTTON | BUTTON_DEFAULT_HIGH, CUSTOM_DEBOUNCE_DELAY, CUSTOM_REPEAT_DELAY);
}

void loop() {
    if (unsigned int event = button->loop()) {
        if (event == EVENT_RELEASED) {
            Serial.print("Count : "); Serial.print(button->getEventCount());
            Serial.print(" Length: "); Serial.print(button->getEventLength());
            Serial.println();
        }
    }
}


examples/debounceevent_switch/debouceevent_switch.cpp
/*

  Debounce buttons and trigger events
  Copyright (C) 2015-2017 by Xose Pérez <xose dot perez at gmail dot com>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <Arduino.h>
#include <DebounceEvent.h>

#define BUTTON_PIN  0

DebounceEvent button = DebounceEvent(BUTTON_PIN, BUTTON_SWITCH | BUTTON_DEFAULT_HIGH);

void setup() {
    Serial.begin(115200);
}

void loop() {
    if (button.loop() != EVENT_NONE) {
        Serial.println("Switched");
    }
}


examples/debounceevent_callback/debouceevent_callback.cpp
/*

  Debounce buttons and trigger events
  Copyright (C) 2015-2017 by Xose Pérez <xose dot perez at gmail dot com>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <Arduino.h>
#include <DebounceEvent.h>

#define BUTTON_PIN     0

void callback(uint8_t pin, uint8_t event, uint8_t count, uint16_t length) {
    Serial.print("Event : "); Serial.print(event);
    Serial.print(" Count : "); Serial.print(count);
    Serial.print(" Length: "); Serial.print(length);
    Serial.println();
}

DebounceEvent button = DebounceEvent(BUTTON_PIN, callback, BUTTON_PUSHBUTTON | BUTTON_DEFAULT_HIGH);

void setup() {
    Serial.begin(115200);
    Serial.println();
    Serial.println();
}

void loop() {
    button.loop();
}


src/DebounceEvent.h
/*

  Debounce buttons and trigger events
  Detects capacitive and resistive touch
  Copyright (C) 2015-2018 by Xose Pérez <xose dot perez at gmail dot com>
  Conttributors: Arjan Mels

  The DebounceEvent library is free software: you can redistribute it and/or modify
  it under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  The DebounceEvent library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with the DebounceEvent library.  If not, see <http://www.gnu.org/licenses/>.


*/

#ifndef _DEBOUNCE_EVENT_h
#define _DEBOUNCE_EVENT_h

#ifdef ESP8266
#include <functional>
#define DEBOUNCE_EVENT_CALLBACK_SIGNATURE std::function<void(uint8_t pin, uint8_t event, uint8_t count, uint16_t length)> callback
#else
#define DEBOUNCE_EVENT_CALLBACK_SIGNATURE void (*callback)(uint8_t pin, uint8_t event, uint8_t count, uint16_t length)
#endif

#define BUTTON_BUTTONTYPE_MASK  (1+8)
#define BUTTON_PUSHBUTTON       0
#define BUTTON_SWITCH           1
#define BUTTON_DEFAULT_HIGH     2
#define BUTTON_SET_PULLUP       4
#define BUTTON_TOUCHBUTTON      8

#define DEBOUNCE_DELAY          50
#define REPEAT_DELAY            500
#define TOUCH_TIMEOUT           1000 // in microseconds

#define TOUCH_DIVIDER           10000   // DIVIDER for teh following alpha (exponential weight values)
#define TOUCH_SHORT_ALPHA       1000    // exponential weight for short term average
#define TOUCH_LONG_ALPHA        1       // exponential weight for long term average 
#define TOUCH_VAR_ALPHA         1       // exponential weight for long term variance 

#define TOUCH_SLOWDOWN          10      // factor to slowdown averaging when touched
#define TOUCH_SENSITIVITY       6       // threshold for detection in x standard deviations from the long term average

#define EVENT_NONE              0
#define EVENT_CHANGED           1
#define EVENT_PRESSED           2
#define EVENT_RELEASED          3

class DebounceEvent {

    public:

        DebounceEvent(uint8_t pin, DEBOUNCE_EVENT_CALLBACK_SIGNATURE, uint8_t mode = BUTTON_PUSHBUTTON | BUTTON_DEFAULT_HIGH, unsigned long delay = DEBOUNCE_DELAY, unsigned long repeat = REPEAT_DELAY, unsigned int touch_timeout = TOUCH_TIMEOUT, unsigned int touch_sensitivity = TOUCH_SENSITIVITY);
        DebounceEvent(uint8_t pin, uint8_t mode = BUTTON_PUSHBUTTON | BUTTON_DEFAULT_HIGH, unsigned long delay = DEBOUNCE_DELAY, unsigned long repeat = REPEAT_DELAY, unsigned int touch_timeout = TOUCH_TIMEOUT, unsigned int touch_sensitivity = TOUCH_SENSITIVITY);
        unsigned char loop();
        bool pressed() { return (_status != _defaultStatus); }
        unsigned long getEventLength() { return _event_length; }
        unsigned long getEventCount() { return _event_count; }
        void setTouchSensitivity(int sensitivity) { _touch_sensitivity = sensitivity; }

    private:

        uint8_t _pin;
        uint8_t _mode;
        unsigned int _touch_timeout;
        bool _status;
        bool _ready = false;
        bool _reset_count = true;
        unsigned long _last_time;
        unsigned long _debounce_start;
        bool _debounce_status;
        unsigned long _event_start;
        unsigned long _event_length;
        unsigned char _event_count;
        bool _defaultStatus;
        unsigned long _delay;
        unsigned long _repeat;
        
        int _touch_alpha;
        unsigned long _touch_short_avg;
        unsigned long _touch_long_avg;
        unsigned long long _touch_long_var;
        bool _touch_prev_touched;
        uint16_t _touch_maxcount; 

        unsigned int _touch_sensitivity; 

        DEBOUNCE_EVENT_CALLBACK_SIGNATURE;

        void _init(uint8_t pin, uint8_t mode, unsigned long delay, unsigned long repeat, unsigned int touch_timeout, unsigned int touch_sensitivity);
        bool _readPin();
        unsigned long _readTouch();
        void _calibrateMaxCount();

};

#endif
